Objects/listobject.c

```diff
@@ -1354,62 +1354,57 @@ gallop_left(MergeState *ms, PyObject *key, PyObject **a, Py_ssize_t n, Py_ssize_
         /* a[hint] < key -- gallop right, until
          * a[hint + lastofs] < key <= a[hint + ofs]
          */
-        const Py_ssize_t maxofs = n - hint;             /* &a[n-1] is highest */
+        const size_t maxofs = n - hint;             /* &a[n-1] is highest */
         while (ofs < maxofs) {
             IFLT(a[ofs], key) {
                 lastofs = ofs;
                 ofs = (ofs << 1) + 1;
-                if (ofs <= 0)                   /* int overflow */
-                    ofs = maxofs;
             }
             else                /* key <= a[hint + ofs] */
                 break;
         }
         if (ofs > maxofs)
             ofs = maxofs;
         /* Translate back to offsets relative to &a[0]. */
-        lastofs += hint;
+        lastofs += hint + 1;
         ofs += hint;
     }
     else {
         /* key <= a[hint] -- gallop left, until
          * a[hint - ofs] < key <= a[hint - lastofs]
          */
-        const Py_ssize_t maxofs = hint + 1;             /* &a[0] is lowest */
+        const size_t maxofs = hint + 1;             /* &a[0] is lowest */
         while (ofs < maxofs) {
             IFLT(*(a-ofs), key)
                 break;
             /* key <= a[hint - ofs] */
             lastofs = ofs;
             ofs = (ofs << 1) + 1;
-            if (ofs <= 0)               /* int overflow */
-                ofs = maxofs;
         }
         if (ofs > maxofs)
             ofs = maxofs;
         /* Translate back to positive offsets relative to &a[0]. */
-        k = lastofs;
-        lastofs = hint - ofs;
-        ofs = hint - k;
+        size_t tmp = lastofs;
+        lastofs = hint + 1 - ofs;
+        ofs = hint - tmp;
     }
     a -= hint;
 
-    assert(-1 <= lastofs && lastofs < ofs && ofs <= n);
-    /* Now a[lastofs] < key <= a[ofs], so key belongs somewhere to the
-     * right of lastofs but no farther right than ofs.  Do a binary
+    assert(lastofs <= ofs && ofs <= (size_t)n);
+    /* Now a[lastofs-1] < key <= a[ofs], so key belongs somewhere to the
+     * right of lastofs-1 but no farther right than ofs.  Do a binary
      * search, with invariant a[lastofs-1] < key <= a[ofs].
```

Comment:
```suggestion
      * search, with invariant a[lastofs] < key <= a[ofs].
```

Reply:
Unlike the first part of the comment, the last line seems to refer to the invariant holding after `++lastofs`, which I moved higher. Therefore, no correction is needed.